openquake/hazardlib/shakemap/conditioned_gmfs.py

```diff
@@ -0,0 +1,903 @@
+        :returns:
+            a 32 bit array of shape (num_imts, num_sites, num_events) and
+            two arrays with shape (num_imts, num_events): sig for tau
+            and eps for the random part
+        """
+        M = len(self.imts)
+        num_sids = mean_covs[0][self.imts[0].string].size
+        result = numpy.zeros((M, num_sids, num_events), F32)
+        sig = numpy.zeros((M, num_events), F32)  # same for all events
+        eps = numpy.zeros((M, num_events), F32)  # not the same
+        numpy.random.seed(self.seed)
+
+        for m, im in enumerate(self.imts):
+            mu_Y_yD = mean_covs[0][im.string]
+            # cov_Y_Y_yD = mean_covs[1][im.string]
+            cov_WY_WY_wD = mean_covs[2][im.string]
+            cov_BY_BY_yD = mean_covs[3][im.string]
+            try:
+                result[m], sig[m], eps[m] = self._compute(
+                    mu_Y_yD, cov_WY_WY_wD, cov_BY_BY_yD, im, num_events, rng)
+            except Exception as exc:
+                raise RuntimeError(
+                    "(%s, %s, source_id=%r) %s: %s"
+                    % (gsim, im, decode(self.source_id),
+                       exc.__class__.__name__, exc)
+                ).with_traceback(exc.__traceback__)
+        if self.amplifier:
+            self.amplifier.amplify_gmfs(
+                self.ctx.ampcode, result, self.imts, self.seed)
+        return result, sig, eps
+
+    def _compute(self, mu_Y, cov_WY_WY, cov_BY_BY, imt, num_events, rng):
+        if self.cmaker.truncation_level <= 1E-9:
+            gmf = exp(mu_Y, imt)  # exponentiate unless imt == 'MMI'
+            gmf = gmf.repeat(num_events, axis=1)
+            inter_sig = 0
+            inter_eps = numpy.zeros(num_events)
+        else:
+            cov_Y_Y = cov_WY_WY + cov_BY_BY
+            gmf = exp(
+                rng.multivariate_normal(
+                    mu_Y.flatten(), cov_Y_Y, size=num_events,
+                    check_valid="warn", tol=1e-5, method="eigh"),
+                imt).T
+            inter_sig = 0
+            inter_eps = 0
+        return gmf, inter_sig, inter_eps  # shapes (N, E), 1, E
+
+
+def get_conditioned_mean_and_covariance(
+        rupture, gmm, station_sitecol, station_data,
+        observed_imt_strs, target_sitecol, target_imts,
+        spatial_correl, cross_correl_between, cross_correl_within,
+        maximum_distance):
+    if hasattr(gmm, "gmpe"):
+        gmm_name = gmm.gmpe.__class__.__name__
+    else:
+        gmm_name = gmm.__class__.__name__
+    if gmm.DEFINED_FOR_STANDARD_DEVIATION_TYPES == {StdDev.TOTAL}:
+        if not (type(gmm).__name__ == "ModifiableGMPE"
+                and "add_between_within_stds" in gmm.kwargs):
+            raise NoInterIntraStdDevs(gmm)
+
+    observed_imtls = {imt_str: [0] for imt_str in observed_imt_strs
+                      if imt_str not in ["MMI", "PGV"]}
+    observed_imts = sorted([imt.from_string(imt_str)
+                            for imt_str in observed_imtls])
```

Comment:
I would like a comment explaing the difference between cmaker_D and cmaker_Y, what is the meaning of the suffixes?

Reply:
Added a comment explaining the difference between cmaker_D and cmaker_Y, and the meaning of the two suffixes